File: cxx_wrapper/LogBuffers.cpp

#include "rust/cxx.h"
#include <memory>
#include "LogBuffers.h"

namespace aeron { namespace logbuffers {
inline std::unique_ptr<aeron::concurrent::AtomicBuffer> atomicBuffer(aeron::LogBuffers & logbuffers, int index){
auto wrapped_buffer = new aeron::concurrent::AtomicBuffer();
    wrapped_buffer->wrap(logbuffers.atomicBuffer(index));
    return std::unique_ptr<AtomicBuffer>(wrapped_buffer);
}

}}

File: src/aeron/log_buffers.rs

#[cxx::bridge(namespace = "aeron")]
pub mod ffi {

    unsafe extern "C++" {
        #[namespace = "aeron::concurrent"]
        type CxxAtomicBuffer = crate::aeron::concurrent::atomic_buffer::ffi::CxxAtomicBuffer;


        include!("aeron-rust-wrapper/aeron/aeron-client/src/main/cpp/LogBuffers.h");
        #[rust_name = "CxxLogBuffers"]
        type LogBuffers;


        include!("aeron-rust-wrapper/cxx_wrapper/LogBuffers.cpp");


        #[namespace = "aeron::logbuffers"]
        #[rust_name = "atomic_buffer"]
        fn atomicBuffer(log_buffers: Pin<&mut CxxLogBuffers>, index: i32) -> UniquePtr<CxxAtomicBuffer>;

    }

    impl SharedPtr<CxxLogBuffers> {}
}